compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +447,243 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
+#endif  // NDEBUG
+      return reifyDynamicResultDims(b, tiedOperand, dynamicDims);
+    }
+  }
+
+  // Case 4: Query ShapeAwareOpInterface.
+  auto shapeAwareOp = dyn_cast<IREE::Util::ShapeAwareOpInterface>(op);
+  if (shapeAwareOp) {
+    ValueRange dims =
+        shapeAwareOp.getResultDynamicDims(opResult.getResultNumber());
+    dynamicDims.append(dims.begin(), dims.end());
+    return success();
+  }
+
+  // Case 5: Query ReifyRankedShapedTypeOpInterface.
+  auto reifyShapeOp = dyn_cast<ReifyRankedShapedTypeOpInterface>(op);
+  if (reifyShapeOp) {
+    ReifiedRankedShapedTypeDims dims;
+    if (failed(reifyShapeOp.reifyResultShapes(b, dims))) return failure();
+    for (int64_t i = 0; i < shapedType.getRank(); ++i)
+      if (shapedType.isDynamicDim(i))
+        dynamicDims.push_back(dims[opResult.getResultNumber()][i]);
+    return success();
+  }
+
+  return failure();
+}
+
+/// Reify the dynamic dimensions of all results of the given op.
+static LogicalResult reifyDynamicResultDims(OpBuilder b, Operation *op,
+                                            SmallVector<Value> &dynamicDims) {
+  for (Value v : op->getResults())
+    if (failed(reifyDynamicResultDims(b, v, dynamicDims))) return failure();
+
+  return success();
+}
+
+static FailureOr<Flow::DispatchRegionOp> wrapInDispatchRegion(
+    RewriterBase &rewriter, Operation *op) {
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(op);
+
+  // Determine dynamic result dims.
+  SmallVector<Value> dynamicDims;
+  if (failed(reifyDynamicResultDims(rewriter, op, dynamicDims)))
+    return failure();
+
+  // Create RegionOp and move op inside.
+  auto regionOp = rewriter.create<Flow::DispatchRegionOp>(
+      op->getLoc(), op->getResultTypes(), dynamicDims);
+  Block &body = regionOp.getBody().emplaceBlock();
+  op->moveBefore(&body, body.begin());
+  op->replaceAllUsesWith(regionOp.getResults());
+
+  // Add terminator.
+  rewriter.setInsertionPointAfter(op);
+  rewriter.create<Flow::ReturnOp>(op->getLoc(), op->getResults());
+
+  return regionOp;
+}
+
+DiagnosedSilenceableFailure
+transform_dialect::WrapInDispatchRegionOp::applyToOne(
+    Operation *target, SmallVectorImpl<Operation *> &results,
+    transform::TransformState &state) {
+  IRRewriter rewriter(target->getContext());
+  FailureOr<Flow::DispatchRegionOp> regionOp =
+      wrapInDispatchRegion(rewriter, target);
+  if (failed(regionOp))
+    return DiagnosedSilenceableFailure(reportUnknownTransformError(target));
+  results.push_back(*regionOp);
+  return DiagnosedSilenceableFailure(success());
+}
+
+static FailureOr<Flow::DispatchRegionOp> moveIntoDispatchRegion(
+    RewriterBase &rewriter, Operation *target,
+    Flow::DispatchRegionOp regionOp) {
+  OpBuilder::InsertionGuard guard(rewriter);
+  Block &body = regionOp.getBody().front();
+  DominanceInfo domInfo;
+
+  // Case 1: `target` dominates `regionOp`.
+  if (domInfo.properlyDominates(target, regionOp)) {
```

Comment:
I am not sure I follow why the consumers of the target are moved into the dispatch region.

Reply:
The `consumers` variable contains all consumers of `target` that are in the dispatch region. In "Case 1" no consumers are moved into the dispatch region. Only in "Case 2", a consumer is moved into the dispatch region.

The semantics of the op is: Given an existing dispatch region, move the target op into the dispatch region. If the target in front of (dominating) the dispatch region, it is a producer (or nothing; in case there is no use of the target in the dispatch region). If the target is after (post-dominating) the dispatch region, it is a consumer (or nothing; in case the target does not use the result of a dispatch region). This op is currently asymmetric in a sense that:
* "dominating but not a producer" --> IR will not be modified
* "post-dominating but not a consumer" --> target will be moved in to the dispatch region,

(Note, Case 1 and Case 2 could be split into two separate ops, but it may be easier to use if the user does not have to think about whether the target is a producer or a consumer.)

Reply:
If this is required by transform dialect, then thats a separate approach. I am not sure building this iteratively this way works very well for a whole program with thousands of statements. The current implementation does this in two steps and makes a distinction between compute ops and clonable ops. Compute ops are not duplicated.

1) So first make a pass through the graph in post-order (use before def). Find the compute ops that become the root operations and then pull in other compute ops that need to be fused with it. I'd expect the same logic to work well for `flow.dispatch.region` op formation. Once you pull something in, it will not even stay outside the graph which means it wont get replicated. Root op determination is simple
- Start with the gemms/convolution, etc. as being the roots. If a gemm is to be fused with its consumer elementwise operation, treat the consumer as the root. (Always easier to pull things into a region from above than to pull things into the region from below).
2) After you form a skeleton of what goes into dispatch region look at values used from outside within the region. For each of these values, walk the use-def chain in reverse recording all operations that can be cloned. Since there are multiple values, you effectively have a DAG of operations. Do a topological sort of them to find order in which they should be cloned and clone them into the region. Now the values used in region defined outside become the actual arguments for the dispatch.

Reply:
Going to what is implemented here. Why "Case 2"? If it is post-dominating, but not a consumer, why is it moved into the dispatch?

Reply:
I dropped "case 2". I still think it's useful to have, but let's leave that for a separate commit to keep this change a bit smaller.

I added an attribute to the op to distinguish between "moving" and "cloning" semantics. Alternatively, I could also make this two separate ops.

At the moment, this op is limited to one target op at a time. I'd like to extend this to multiple targets (and the topological sorting) in a separate commit if that's OK.